modules/setting/setting.go

```diff
@@ -954,6 +956,20 @@ func loadFromConf(allowEmpty bool, extraConfig string) {
 	PasswordCheckPwn = sec.Key("PASSWORD_CHECK_PWN").MustBool(false)
 	SuccessfulTokensCacheSize = sec.Key("SUCCESSFUL_TOKENS_CACHE_SIZE").MustInt(20)
 
+	// Master key provider configuration
+	MasterKeyProvider = sec.Key("MASTER_KEY_PROVIDER").MustString("plain")
+	switch MasterKeyProvider {
+	case "plain":
+		MasterKey = []byte(sec.Key("MASTER_KEY").MustString(SecretKey))
+		if len(MasterKey) > 32 {
+			MasterKey = MasterKey[:32]
+		}
```

Comment:
We should not do this. Just accept all lengths and use a key derivation function to transform a password into the needed size.

https://pkg.go.dev/golang.org/x/crypto@v0.0.0-20221012134737-56aed061732a/pbkdf2

Reply:
It needs a random salt, where should we store it?

Reply:
I'm not an expert but we might simply pass a static salt (`gitea`) because we only use this method to transform the user password to be usable by AES. If someone steals the ini with the secret, the stored salt is leaked too. So we could just use a static string.

The RFC states two benefits for the salt:
1. It is difficult for an opponent to precompute all the keys corresponding to a dictionary of passwords, or even the most likely keys.
2. It is unlikely that the same key will be selected twice.

Both doesn't apply for us.

Reply:
Done.